Expensify/App
# fixed messages are repeated on page refresh & process queue only when client is leader

Please review @marcaaron.



### Details

1. This PR tries to fix the duplicate request issue from the Web platform. 
2. On the web, a user can open multiple tabs for E.cash. And all of these clients are able to send requests. 
3. But the User can go offline any time. we have a mechanism to handle requests when the user is offline.
4. Currently there is an issue that these persisted requests are duplicated when the user comes back online. 
5. Each client(tab) has an in-memory network queue. and We have global NETWORK_REQUEST_QUEUE for persisted requests.
6. When a user is offline all the clients will push their requests to an in-memory network queue, from which we filter all the retriable requests and push those to our persisted queue. We filter the in-memory queue and remove these persisted requests that are stored in the persisted queue.
7. Due to Onyx, this persisted queue is synced across the clients.
8. When the user comes back online, we ask the leader client to process all the persisted requests from each client. Thus no request will be duplicated. After all the requests are sent to the server, messages will be sync across tabs via Onyx, keeping the same report actions for all clients.

### Fixed Issues

Fixes #2451
Fixes #4214

### Tests / QA Steps
1. Log in to staging.expensify.cash
2. Navigate to a conversation
3. Disable the internet connection
4. Send a message while offline
5. Enable the internet connection and immediately refresh the page.
6. Messages should not be repeated.

### Tested On

- [x] Web (Affected Platfrom)

### Screenshots


#### Web
 

<details>
<summary><h5 style="display:inline">Before</h5></summary>

https://user-images.githubusercontent.com/44479856/115265297-18540500-a105-11eb-8019-d284211a6b0b.mp4

</details>
<details open>
<summary><h3 style="display:inline">After</h3></summary>

https://user-images.githubusercontent.com/24370807/117200247-e7fcae00-ae08-11eb-8284-6ef0404129d6.mp4


</details>
